examples_utils/benchmarks/run_benchmarks.py

```diff
@@ -127,39 +127,50 @@ def run_and_monitor_progress(
     def kill_process(proc_pid: int):
         process = psutil.Process(proc_pid)
         for proc in process.children(recursive=True):
-            logger.info("Killing child process %s" % proc.pid)
+            logger.info("Killing child process %s", proc.pid)
             proc.kill()
-        logger.info("Killing process %s" % proc_pid)
+        logger.info("Killing process %s", proc_pid)
         process.kill()
 
     def proc_thread():
         sel = selectors.DefaultSelector()
         sel.register(proc.stdout, selectors.EVENT_READ)
         sel.register(proc.stderr, selectors.EVENT_READ)
         eof = False
+        decode_error_count = 0
         while not eof:
-            for key, _ in sel.select():
+            selected = 0
+            for key, _ in sel.select(timeout=10):
+                selected += 1
                 stream = key.fileobj
                 data = stream.read1(80)
-                try:
-                    data = data.decode()
-                    if not data:
-                        eof = True
-                    listener.write(data)
-                    listener.flush()
-
-                    if stream is proc.stdout:
-                        outs[0].append(data)
-                    else:
-                        outs[1].append(data)
-                except UnicodeDecodeError as e:
-                    pass
-
-        out, err = proc.communicate()
-        outs[0].append(out.decode())
-        listener.write(out.decode())
-        outs[1].append(err.decode())
-        listener.write(err.decode())
+                data = data.decode(errors="backslashreplace")
+                if not data:
+                    eof = True
+                listener.write(data)
+                listener.flush()
+
+                if stream is proc.stdout:
+                    outs[0].append(data)
+                else:
+                    outs[1].append(data)
+            if not selected:
+                logger.debug("Selector did not pick any files to explore, polling to check for exit")
+                if proc.poll() is not None:
```

Comment:
Is it possible for `if proc.poll() is not None:` to return `False`, as then I think we’ll still be stuck in an infinite loop?

Reply:
We actually often hit `False` on this condition, the `select` timeout is often reached even as the underlying process is running correctly. In an earlier version of this PR I had this line:

```
logger.info("Selector did not pick any files to explore, polling to check for exit")
```

 and I saw it appear in the logs very often.

Reply:
Is it possible/likely for the subprocess to hang and `proc.poll()` return `None` and we get stuck in an infinite loop because of that?

Reply:
As you say, that is handled by the other timeout